test/affinity.cxx

```diff
@@ -0,0 +1,430 @@
+}
+
+static void nextProcessor(const processorVector_t &processorInfo,
+	processorIterator_t &info, uint16_t &groupIndex, uint8_t &maskOffset)
+{
+	while (info != processorInfo.end())
+	{
+		while (info->Relationship != RelationGroup)
+		{
+			++info;
+			REQUIRE(info != processorInfo.end());
+		}
+		auto processor{*info};
+		REQUIRE(groupIndex < processor.Group.ActiveGroupCount);
+		for (; groupIndex < processor.Group.ActiveGroupCount; ++groupIndex)
+		{
+			const auto &group{processor.Group.GroupInfo[groupIndex]};
+			for (; maskOffset < sizeof(KAFFINITY) * 8; ++maskOffset)
+			{
+				const auto mask{group.ActiveProcessorMask >> maskOffset};
+				if (mask & 1)
+					return;
+			}
+			maskOffset = 0;
+		}
+		++info;
+	}
+	throw std::runtime_error("Could not find another valid group:processor combo while there are entries left in affinity_t");
+}
+#elif defined(__APPLE__)
+static uint32_t sched_getcpu()
+{
+	boolean_t getDefault{false};
+	thread_affinity_policy_data_t policy{};
+	mach_msg_type_number_t numPolicies{1};
+	const thread_port_t machThread{mach_thread_self()};
+	// NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+	REQUIRE(thread_policy_get(machThread, THREAD_AFFINITY_POLICY, reinterpret_cast<thread_policy_t>(&policy), &numPolicies, &getDefault) == 0);
+	return static_cast<uint32_t>(policy.affinity_tag);
+};
+#endif
+
+TEST_CASE("construct", "[affinity_t]")
+{
+	const auto affinity{substrate::make_unique_nothrow<substrate::affinity_t>()};
+	REQUIRE(affinity);
+}
+
+TEST_CASE("processor count", "[affinity_t]")
+{
+	const auto affinity{substrate::make_unique_nothrow<substrate::affinity_t>()};
+	REQUIRE(affinity);
+	REQUIRE(affinity->numProcessors() == std::thread::hardware_concurrency());
+}
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
+TEST_CASE("iteration", "[affinity_t]")
+{
+	const auto affinity{substrate::make_unique_nothrow<substrate::affinity_t>()};
+	REQUIRE(affinity);
+#if defined(_POSIX_THREADS) && defined(_GNU_SOURCE)
+	cpu_set_t affinitySet{};
+	REQUIRE(sched_getaffinity(0, sizeof(cpu_set_t), &affinitySet) == 0);
+
+	REQUIRE(affinity->begin() != affinity->end());
+	std::size_t count{0};
+	for (const auto processor : *affinity)
+	{
+		++count;
+		REQUIRE(CPU_ISSET(processor, &affinitySet));
+	}
+	REQUIRE(count == affinity->numProcessors());
+#elif defined(_POSIX_THREADS) && defined(__APPLE__)
+	boolean_t getDefault{false};
+	thread_affinity_policy_data_t policy{};
+	mach_msg_type_number_t numPolicies{1};
+	const thread_port_t machThread{mach_thread_self()};
+	// NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+	REQUIRE(thread_policy_get(machThread, THREAD_AFFINITY_POLICY, reinterpret_cast<thread_policy_t>(&policy), &numPolicies, &getDefault) == 0);
+	REQUIRE(affinity->begin() != affinity->end());
+	std::size_t count{0};
+	for (const auto processor: *affinity)
+	{
+		(void)processor;
+		++count;
+		REQUIRE(((policy.affinity_tag & 1 << count) || !policy.affinity_tag));
```

Comment:
We're not completely clear on the evaluation order that occurs here, so could you put ()'s around the part that should happen first please - eg, `(policy.afinity_tag & (1 << count))`